### PR TITLE
Test against Pandas nightly

### LINKED ISSUE #11: Nightly builds
We will depend on Pandas and Apache Arrow master for a while. They will also change from time to time and breakage is expected. Once we have CI up and running, we should also check continuously that upstream introduces no new breakages.

### LINKED ISSUE #43: List Type testing
Add a `list` type to the unit tests and fix the failing ones.

### LINKED ISSUE #44: Dtypes names should be "fletcher[X]"
Otherwise they look like they are official Pandas types.

### LINKED ISSUE #46: FletcherArray.__setitem__(ndarray, FletcherArray) ignores null bitmap
`series[series.isna()] = series[series.isna()]` remove all null entries and replaces them with garbage.

### PR DESCRIPTION
**Pull request by xhochy**
Adds tests for list, int64 and float64 types. Sadly they all fail on all pandas releases. I have marked the failures and will fix them upstream.

Fixes #43 
Fixes #44 
Fixes #46 
Fixes #11

### CODE REVIEW

File: fletcher/base.py [hunk 1]
```diff
@@ -442,6 +481,8 @@ def astype(self, dtype, copy=True):
             raise NotImplementedError("Cast propagation in astype not yet implemented")
         else:
             dtype = np.dtype(dtype)
+            if pa.types.is_list(self.dtype.arrow_dtype) and dtype.kind == "U":
```
Comment (fjetter) [inline fletcher/base.py]: Why only for Unicode?

File: tests/test_pandas_extension.py [hunk 2]
```diff
@@ -29,9 +33,19 @@
         "data_for_grouping",
         "data_for_sorting",
         "data_missing_for_sorting",
+        "data_repeated",
     ],
 )
 
+if sys.version_info >= (3, 6):
```
Comment (fjetter) [inline tests/test_pandas_extension.py]: Six?

File: tests/test_pandas_extension.py [hunk 3]
```diff
@@ -169,6 +229,36 @@ def test_reindex(self):
         # due to a dtype assumption that does not hold for Arrow
         pass
 
+    def test_get(self, data, dtype):
+        if pa.types.is_list(dtype.arrow_dtype):
+            pytest.skip("assert_extension_array_equal casts to numpy object arrays")
```
Comment (fjetter) [inline tests/test_pandas_extension.py]: Maybe use xfail in case the tests are ever rewritten we'd notice

File: tests/test_pandas_extension.py [hunk 4]
```diff
@@ -179,15 +269,65 @@ class TestBaseInterfaceTests(BaseInterfaceTests):
 
 
 class TestBaseMethodsTests(BaseMethodsTests):
-    pass
+
+    @pytest.mark.parametrize("dropna", [True, False])
+    def test_value_counts(self, all_data, dropna, dtype):
+        # Skip integer tests while there is no support for ExtensionIndex.
```
Comment (fjetter) [inline tests/test_pandas_extension.py]: What does the index have to do with this?

File: tests/test_pandas_extension.py [hunk 5]
```diff
@@ -179,15 +269,65 @@ class TestBaseInterfaceTests(BaseInterfaceTests):
...
 class TestBaseMissingTests(BaseMissingTests):
     pass
 
 
 class TestBaseReshapingTests(BaseReshapingTests):
-    pass
+
+    def test_concat_mixed_dtypes(self, data, dtype):
+        if dtype.name in ["fletcher[int64]", "fletcher[double]"]:
+            # TODO: Raise issue if this is expected.
```
Comment (fjetter) [inline tests/test_pandas_extension.py]: Have you opened the ticket, yet?

**Comment by fjetter** [comment]
Maybe we can "catch" the failing test s.t. We can still look up how it is performing but it doesn't interfere with the notifications.

File: tests/test_pandas_extension.py (hunk 2 above)
Comment (xhochy) [reply tests/test_pandas_extension.py]: No, six only distinguishes between Python 2 and 3. This is a 3.6+ feature that is not available in 3.5.

Comment (fjetter) [reply tests/test_pandas_extension.py]: You're right. I thought I saw six variables to distinguish between py3 versions. I was wrong, though.

File: fletcher/base.py [hunk 6]
```diff
@@ -261,8 +283,18 @@ def __setitem__(self, key, value):
...
+                    # In the case where we zero-copy Arrow to Pandas conversion, the
+                    # the resulting arrays are read-only.
+                    if not arr.flags.writeable:
+                        arr = arr.copy()
                     arr[array_chunk_indices] = np.array(value)[key_chunk_indices]
-                    chunks.append(pa.array(arr, self.dtype.arrow_dtype))
+                    if pa.types.is_integer(
+                        self.dtype.arrow_dtype
+                    ) or pa.types.is_floating(self.dtype.arrow_dtype):
+                        mask = pd.isna(arr)
```
[the remark this answers is not in the document]
Comment (xhochy) [reply fletcher/base.py]: Opened https://issues.apache.org/jira/browse/ARROW-2806 about this. Will mention as a comment.

File: fletcher/base.py (hunk 1 above)
Comment (xhochy) [reply fletcher/base.py]: NumPy has implemented a different conversion for `list->unicode` than the default Python `__str__` method would do, the expectation is that the Python version is used.

File: tests/test_pandas_extension.py (hunk 3 above)
Comment (xhochy) [reply tests/test_pandas_extension.py]: As there is a general `xfail` for list types, I have removed all the test specializations.

**Pull request merged**